dbms/src/DataStreams/RuntimeFilter.cpp

```diff
@@ -0,0 +1,217 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <DataStreams/RuntimeFilter.h>
+#include <Interpreters/Set.h>
+#include <Storages/DeltaMerge/FilterParser/FilterParser.h>
+#include <common/logger_useful.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int SET_SIZE_LIMIT_EXCEEDED;
+}
+
+std::string RuntimeFilter::getSourceColumnName() const
+{
+    return source_column_name;
+}
+
+tipb::RuntimeFilterType RuntimeFilter::getRFType() const
+{
+    return rf_type;
+}
+
+tipb::Expr RuntimeFilter::getSourceExpr() const
+{
+    return source_expr;
+}
+
+std::string RuntimeFilter::getFailedReason()
+{
+    std::lock_guard<std::mutex> lock(mtx);
+    return failed_reason;
+}
+
+void RuntimeFilter::setSourceColumnName(const std::string & source_column_name_)
+{
+    source_column_name = source_column_name_;
+}
+
+void RuntimeFilter::setINValuesSet(const std::shared_ptr<Set> & in_values_set_)
+{
+    in_values_set = in_values_set_;
+}
+
+void RuntimeFilter::build()
+{
+    if (!DM::FilterParser::isRSFilterSupportType(target_expr.field_type().tp()))
+    {
+        throw TiFlashException(Errors::Coprocessor::Unimplemented,
+                               "The rs operator doesn't support target expr type:{}, rf_id:{}",
+                               target_expr.field_type().DebugString(),
+                               id);
+    }
+    if (source_expr.tp() != tipb::ExprType::ColumnRef)
+    {
+        throw TiFlashException(Errors::Coprocessor::BadRequest,
+                               "The source expr {} of rf {} should be column ref",
+                               source_expr.tp(),
+                               id);
+    }
+}
+
+// There is a possibility that the same rf is canceled multiple times,
+// but rf does not have thread safety issues.
+void RuntimeFilter::updateValues(const ColumnWithTypeAndName & values, const LoggerPtr & log)
+{
+    if (isFailed())
+    {
+        return;
+    }
+    Block data(std::initializer_list<ColumnWithTypeAndName>({values}));
+    switch (rf_type)
+    {
+    case tipb::IN:
+        try
+        {
+            LOG_DEBUG(log, "update values in rf, data size:{}", data.rows());
+            in_values_set->insertFromBlock(data, false);
+        }
+        catch (Exception & e)
+        {
+            if (e.code() == ErrorCodes::SET_SIZE_LIMIT_EXCEEDED)
+            {
+                std::string tmp_err_msg = "The rf in values exceed the limit. " + e.message();
+                updateStatus(RuntimeFilterStatus::FAILED, tmp_err_msg);
+                LOG_WARNING(log, "cancel runtime filter id:{}, reason: {} ", id, tmp_err_msg);
+            }
+        }
+        break;
+    case tipb::MIN_MAX:
+    case tipb::BLOOM_FILTER:
+        // todo
+        break;
+    }
+}
+
+void RuntimeFilter::finalize(const LoggerPtr & log)
+{
+    if (updateStatus(RuntimeFilterStatus::READY))
```

Comment:
should be `if (!updateStatus(RuntimeFilterStatus::READY))` ?

Reply:
fixed